mPowerSDK/APHScoring.m

```diff
@@ -87,20 +104,287 @@ - (NSDictionary *)summarizeDataset:(NSDictionary *)dataset period:(APHTimelineGr
+    self.medTimingDataPointsCorrelatedScoring = [[NSMutableDictionary alloc] init];
+    
+	APHMedicationTrackerTask *task = [[APHMedicationTrackerTask alloc] init];
+	NSArray<ORKTextChoice *> *activityTimingChoices = task.activityTimingChoices;
+    
+    if (!self.filteredDataPoints) {
+        self.filteredDataPoints = [NSArray arrayWithArray:self.dataPoints.copy];
+    }
+
+	for (ORKTextChoice *textChoice in activityTimingChoices) {
+		self.medTimingDataPoints[textChoice.text] = [[NSMutableArray alloc] init];
+	}
+    
+    NSString *noMedicationTimingKey = self.activityTimingChoicesStrings.lastObject;
+
+	for (NSDictionary *dataPoint in [self.dataPoints copy]) {
+        NSArray *rawDataPoints = dataPoint[@"datasetRawDataPoints"];
+        
+        for (NSString *activityTimingChoiceString in self.activityTimingChoicesStrings) {
+            NSMutableDictionary *filteredDataPoint = [dataPoint mutableCopy];
+            
+            NSPredicate *filterPredicate;
+            if ([activityTimingChoiceString isEqualToString:noMedicationTimingKey]) {
+                filterPredicate = [NSPredicate predicateWithFormat:@"(datasetTaskResult == nil) || (datasetTaskResult.MedicationActivityTiming == nil)"];
+            } else {
+                filterPredicate = [NSPredicate predicateWithFormat:@"(datasetTaskResult.MedicationActivityTiming == %@)", activityTimingChoiceString];
+            }
+            
+            NSArray *filteredRawDataPoints = [rawDataPoints filteredArrayUsingPredicate:filterPredicate];
+            filteredDataPoint[@"datasetRawDataPoints"] = filteredRawDataPoints;
+            
+            if (filteredRawDataPoints.count == 0 && ![activityTimingChoiceString isEqualToString:noMedicationTimingKey]) {
+                filteredDataPoint[kDatasetValueKey] = @(NSNotFound);
+            } else if (filteredRawDataPoints.count > 0) {
+                NSNumber *averageRawDataPointValue = [[filteredRawDataPoints valueForKey:kDatasetValueKey] valueForKeyPath:@"@avg.intValue"];
+                filteredDataPoint[kDatasetValueKey] = averageRawDataPointValue;
+            }
+            
+            [self.medTimingDataPoints[activityTimingChoiceString] addObject:filteredDataPoint];
+        }
+	}
+    
+    NSMutableArray *averagedDataPoints = [NSMutableArray new];
+    for (NSDictionary *dataPoint in [self.dataPoints copy]) {
+        NSMutableDictionary *copiedDataPoint = [dataPoint mutableCopy];
+        
+        NSMutableArray *newRawDataPoints = [NSMutableArray new];
+        for (NSString *medTimingChoiceString in self.activityTimingChoicesStrings) {
+            NSArray *points = self.medTimingDataPoints[medTimingChoiceString];
+            NSPredicate *filterPredicate = [NSPredicate predicateWithFormat:@"datasetDateKey == %@", dataPoint[kDatasetDateKey]];
+            NSDictionary *point = [points filteredArrayUsingPredicate:filterPredicate].firstObject;
+            
+            if (!point) {
```

Comment:
Ugh, please don't. Just wrap the code where you do something with the point in an if statement `if (point != nil)`. Early exit has it's advantages, but not in this case.